donkeycar/parts/pins.py

```diff
@@ -0,0 +1,913 @@
+    gpio_pin_pull = [None, GPIO.PUD_OFF, GPIO.PUD_DOWN, GPIO.PUD_UP]
+    gpio_pin_scheme = {PinScheme.BOARD: GPIO.BOARD, PinScheme.BCM: GPIO.BCM}
+except ImportError:
+    print("pigpio was not imported.")
+    globals()["GPIO"] = None
+
+
+def gpio_fn(pin_scheme, fn):
+    """
+    Convenience method to call GPIO function
+    using desired pin scheme.  This restores
+    the previous pin scheme, so it is safe to 
+    mix pin schemes.
+    """
+    prev_scheme = GPIO.getmode() or pin_scheme
+    GPIO.setmode(pin_scheme)
+    val = fn()
+    GPIO.setmode(prev_scheme)
+    return val
+
+
+class InputPinGpio(InputPin):
+    def __init__(self, pin_number:int, pin_scheme:str, pull=PinPull.PULL_NONE) -> None:
+        """
+        Input pin ttl HIGH/LOW using RPi.GPIO/Jetson.GPIO
+        pin_number: GPIO.BOARD mode point number
+        pull: enable a pull up or down resistor on pin.  Default is PinPull.PULL_NONE
+        """
+        self.pin_number = pin_number
+        self.pin_scheme = gpio_pin_scheme[pin_scheme]
+        self.pull = pull
+        self.on_input = None
+        self._state = PinState.NOT_STARTED
+        super().__init__()
+
+    def _callback(self, pin_number):
+        if self.on_input is not None:
+            self.on_input(self.pin_number, self.input())
+
+    def start(self, on_input=None, edge=PinEdge.RISING) -> None:
+        """
+        on_input: function to call when an edge is detected, or None to ignore
+        edge: type of edge(s) that trigger on_input; default is 
+        """
+        if self.state() != PinState.NOT_STARTED:
+            raise RuntimeError("Attempt to start InputPin that is already started.")
+        gpio_fn(self.pin_scheme, lambda: GPIO.setup(self.pin_number, GPIO.IN, pull_up_down=gpio_pin_pull[self.pull]))
+        if on_input is not None:
+            self.on_input = on_input
+            gpio_fn(self.pin_scheme, lambda: GPIO.add_event_detect(self.pin_number, gpio_pin_edge[edge], callback=self._callback))
+        self.input()  # read first state
+
+    def stop(self) -> None:
+        if self.state() != PinState.NOT_STARTED:
+            gpio_fn(self.pin_scheme, lambda: GPIO.cleanup(self.pin_number))
+            self._state = PinState.NOT_STARTED
+
+    def state(self) -> int:
+        return self._state
+
+    def input(self) -> int:
+        self._state = gpio_fn(self.pin_scheme, lambda: GPIO.input(self.pin_number))
+        return self._state
+
+
+class OutputPinGpio(OutputPin):
+    """
+    Output pin ttl HIGH/LOW using Rpi.GPIO/Jetson.GPIO
+    """
+    def __init__(self, pin_number:int, pin_scheme:str) -> None:
+        self.pin_number = pin_number
+        self.pin_scheme = gpio_pin_scheme[pin_scheme]
+        self._state = PinState.NOT_STARTED
+
+    def start(self, state:int=PinState.LOW) -> None:
+        if self.state() != PinState.NOT_STARTED:
+            raise RuntimeError("Attempt to start OutputPin that is already started.")
```

Comment:
Maybe add the PIN number to the exception message, so we know which ones is causing a problem.

Reply:
Done

Reply:
...? not in this version.. maybe still on your local change...